File: SmartDeviceLink_Example/VehicleDataManager.m
//
//  VehicleDataManager.m
//  SmartDeviceLink-Example-ObjC
//
//  Created by Nicole on 4/23/18.
//  Copyright © 2018 smartdevicelink. All rights reserved.
//

#import "AlertManager.h"
#import "VehicleDataManager.h"
#import "AppConstants.h"
#import "SmartDeviceLink.h"

NS_ASSUME_NONNULL_BEGIN


@interface VehicleDataManager ()

@property (strong, nonatomic) SDLManager *sdlManager;
@property (copy, nonatomic) NSString *vehicleOdometerData;
@property (copy, nonatomic, nullable) RefreshUIHandler refreshUIHandler;

@end

@implementation VehicleDataManager

#pragma mark - Lifecycle

- (instancetype)initWithManager:(SDLManager *)manager refreshUIHandler:(RefreshUIHandler)refreshUIHandler {
    self = [super init];
    if (!self) {
        return nil;
    }

    _sdlManager = manager;
    _refreshUIHandler = refreshUIHandler;
    _vehicleOdometerData = @"";

    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(vehicleDataNotification:) name:SDLDidReceiveVehicleDataNotification object:nil];
    [self sdlex_resetOdometer];

    return self;
}

- (void)stopManager {
    [self sdlex_resetOdometer];
}

#pragma mark - Subscribe Vehicle Data

- (void)subscribeToVehicleOdometer {
    SDLLogD(@"Subscribing to odometer vehicle data");
    SDLSubscribeVehicleData *subscribeToVehicleOdometer = [[SDLSubscribeVehicleData alloc] init];
    subscribeToVehicleOdometer.odometer = @YES;
    [self.sdlManager sendRequest:subscribeToVehicleOdometer withResponseHandler:^(__kindof SDLRPCRequest * _Nullable request, __kindof SDLRPCResponse * _Nullable response, NSError * _Nullable error) {
        if (error || ![response isKindOfClass:SDLGetVehicleDataResponse.class]) {
            SDLLogD(@"Error sending Get Vehicle Data RPC: %@", error);
            return;
        }

        SDLGetVehicleDataResponse* getVehicleDataResponse = (SDLGetVehicleDataResponse *)response;
        SDLResult resultCode = getVehicleDataResponse.resultCode;

        NSMutableString *message = [NSMutableString stringWithFormat:@"%@: ", VehicleDataOdometerName];
        if ([resultCode isEqualToEnum:SDLResultSuccess]) {
            SDLLogD(@"Subscribed to vehicle odometer data");
            [message appendString:@"Subscribed"];
        } else if ([resultCode isEqualToEnum:SDLResultDisallowed]) {
            SDLLogD(@"Access to vehicle data disallowed");
            [message appendString:@"Disallowed"];
        } else if ([resultCode isEqualToEnum:SDLResultUserDisallowed]) {
            SDLLogD(@"Vehicle user disabled access to vehicle data");
            [message appendString:@"Disabled"];
        } else if ([resultCode isEqualToEnum:SDLResultIgnored]) {
            SDLLogD(@"Already subscribed to odometer data");
            [message appendString:@"Subscribed"];
        } else if ([resultCode isEqualToEnum:SDLResultDataNotAvailable]) {
            SDLLogD(@"You have permission to access to vehicle data, but the vehicle you are connected to did not provide any data");
            [message appendString:@"Unknown"];
        } else {
            SDLLogD(@"Unknown reason for failure to get vehicle data: %@", error != nil ? error.localizedDescription : @"no error message");
            [message appendString:@"Unsubscribed"];
        }

        self.vehicleOdometerData = message;

        if (!self.refreshUIHandler) { return; }
        self.refreshUIHandler();
    }];
}

- (void)unsubscribeToVehicleOdometer {
    SDLUnsubscribeVehicleData *unsubscribeToVehicleOdometer = [[SDLUnsubscribeVehicleData alloc] init];
    unsubscribeToVehicleOdometer.odometer = @YES;
    [self.sdlManager sendRequest:unsubscribeToVehicleOdometer withResponseHandler:^(__kindof SDLRPCRequest * _Nullable request, __kindof SDLRPCResponse * _Nullable response, NSError * _Nullable error) {
        if (!response.success.boolValue) { return; }
        [self sdlex_resetOdometer];
    }];
}


- (void)vehicleDataNotification:(SDLRPCNotificationNotification *)notification {
    if (![notification.notification isKindOfClass:SDLOnVehicleData.class]) {
        return;
    }

    SDLOnVehicleData *onVehicleData = (SDLOnVehicleData *)notification.notification;
    self.vehicleOdometerData = [NSString stringWithFormat:@"%@ = %@ kph", VehicleDataOdometerName, onVehicleData.odometer];

    if (!self.refreshUIHandler) { return; }
    self.refreshUIHandler();
}

- (void)sdlex_resetOdometer {
    self.vehicleOdometerData = [NSString stringWithFormat:@"%@ = %@ kph", VehicleDataOdometerName, self.vehicleOdometerData];
}

#pragma mark - Get Vehicle Data

+ (void)getVehicleSpeedWithManager:(SDLManager *)manager {
    SDLLogD(@"Checking if app has permission to access vehicle data...");
    if (![manager.permissionManager isRPCAllowed:@"GetVehicleData"]) {
        [manager sendRequest:[AlertManager alertWithMessageAndCloseButton:@"This app does not have the required permissions to access vehicle data" textField2:nil]];
        return;
    }

    SDLLogD(@"App has permission to access vehicle data. Requesting vehicle speed data...");
    SDLGetVehicleData *getVehicleSpeed = [[SDLGetVehicleData alloc] init];
    getVehicleSpeed.speed = @YES;
    [manager sendRequest:getVehicleSpeed withResponseHandler:^(__kindof SDLRPCRequest * _Nullable request, __kindof SDLRPCResponse * _Nullable response, NSError * _Nullable error) {
        if (error || ![response isKindOfClass:SDLGetVehicleDataResponse.class]) {
            [manager sendRequest:[AlertManager alertWithMessageAndCloseButton:@"Something went wrong while getting vehicle speed" textField2:nil]];
            return;
        }

        SDLGetVehicleDataResponse* getVehicleDataResponse = (SDLGetVehicleDataResponse *)response;
        SDLResult resultCode = getVehicleDataResponse.resultCode;

        NSMutableString *alertMessage = [NSMutableString stringWithFormat:@"%@: ", VehicleDataSpeedName];
        if ([resultCode isEqualToEnum:SDLResultRejected]) {
            SDLLogD(@"The request for vehicle speed was rejected");
            [alertMessage appendString:@"Rejected"];
        } else if ([resultCode isEqualToEnum:SDLResultDisallowed]) {
            SDLLogD(@"This app does not have the required permissions to access vehicle data.");
            [alertMessage appendString:@"Disallowed"];
        } else if ([resultCode isEqualToEnum:SDLResultSuccess]) {
            NSNumber *speed = getVehicleDataResponse.speed;
            if (speed) {
                SDLLogD(@"Request for vehicle speed successful: %f", speed.floatValue);
                [alertMessage appendString:[NSString stringWithFormat:@"%f kph", speed.floatValue]];
            } else {
                SDLLogD(@"Request for vehicle speed successful but no data returned.");
                [alertMessage appendString:@"Unknown"];
            }
        }

        [manager sendRequest:[AlertManager alertWithMessageAndCloseButton:alertMessage textField2:nil]];
    }];
}

#pragma mark - Phone Calls

+ (void)checkPhoneCallCapabilityWithManager:(SDLManager *)manager phoneNumber:(NSString *)phoneNumber {
    SDLLogD(@"Checking phone call capability");
    [manager.systemCapabilityManager updateCapabilityType:SDLSystemCapabilityTypePhoneCall completionHandler:^(NSError * _Nullable error, SDLSystemCapabilityManager * _Nonnull systemCapabilityManager) {
        if (!systemCapabilityManager.phoneCapability) {
            [manager sendRequest:[AlertManager alertWithMessageAndCloseButton:@"The head unit does not support the phone call  capability" textField2:nil]];
            return;
        }

        if (systemCapabilityManager.phoneCapability.dialNumberEnabled.boolValue) {
            SDLLogD(@"Dialing phone number %@", phoneNumber);
            [self sdlex_dialPhoneNumberWithManager:manager phoneNumber:phoneNumber];
        } else {
            [manager sendRequest:[AlertManager alertWithMessageAndCloseButton:@"The dial number feature is unavailable for this head unit" textField2:nil]];
        }
    }];
}

+ (void)sdlex_dialPhoneNumberWithManager:(SDLManager *)manager phoneNumber:(NSString *)phoneNumber {
    SDLDialNumber *dialNumber = [[SDLDialNumber alloc] initWithNumber:phoneNumber];
    [manager sendRequest:dialNumber withResponseHandler:^(__kindof SDLRPCRequest * _Nullable request, __kindof SDLRPCResponse * _Nullable response, NSError * _Nullable error) {
        if (!response.resultCode) { return; }
        SDLLogD(@"Sent dial number request: %@", response.resultCode == SDLResultSuccess ? @"successfully" : @"unsuccessfully");
    }];
}

@end

NS_ASSUME_NONNULL_END

File: SmartDeviceLink_Example/VehicleDataManager.swift
//
//  VehicleDataManager.swift
//  SmartDeviceLink
//
//  Created by Nicole on 4/13/18.
//  Copyright © 2018 smartdevicelink. All rights reserved.
//

import Foundation
import SmartDeviceLink
import SmartDeviceLinkSwift

enum VehicleData: String {
    case accPedalPosition = "APP"
    case speed = "Speed"
    case gps = "GPS"
}

class VehicleDataManager: NSObject {
    fileprivate let sdlManager: SDLManager!
    fileprivate var refreshUIHandler: RefreshUIHandler?
    public fileprivate(set) var vehicleOdometerData: String

    init(sdlManager: SDLManager, refreshUIHandler: RefreshUIHandler? = nil) {
        self.sdlManager = sdlManager
        self.refreshUIHandler = refreshUIHandler
        self.vehicleOdometerData = ""
        super.init()

        resetOdometer()
        NotificationCenter.default.addObserver(self, selector: #selector(vehicleDataNotification(_:)), name: .SDLDidReceiveVehicleData, object: nil)
    }
}

// MARK: - Subscribe Vehicle Data

extension VehicleDataManager {
    /// Subscribes to odometer data. You must subscribe to a notification with name `SDLDidReceiveVehicleData` to get the new data when the odometer data changes.
    func subscribeToVehicleOdometer() {
        let subscribeToVehicleOdometer = SDLSubscribeVehicleData()
        subscribeToVehicleOdometer.odometer = true
        sdlManager.send(request: subscribeToVehicleOdometer) { [unowned self] (request, response, error) in
            guard let result = response?.resultCode else { return }

            if error != nil {
                SDLLog.e("Error sending Get Vehicle Data RPC: \(error!.localizedDescription)")
            }

            var message = "\(VehicleDataOdometerName): "
            switch result {
            case .success:
                SDLLog.d("Subscribed to vehicle odometer data")
                message += "Subscribed"
            case .disallowed:
                SDLLog.d("Access to vehicle data disallowed")
                message += "Disallowed"
            case .userDisallowed:
                SDLLog.d("Vehicle user disabled access to vehicle data")
                message += "Disabled"
            case .ignored:
                SDLLog.d("Already subscribed to odometer data")
                message += "Subscribed"
            case .dataNotAvailable:
                SDLLog.d("You have permission to access to vehicle data, but the vehicle you are connected to did not provide any data")
                message += "Unknown"
            default:
                SDLLog.e("Unknown reason for failure to get vehicle data: \(error != nil ? error!.localizedDescription : "no error message")")
                message += "Unsubscribed"
                return
            }
            self.vehicleOdometerData = message

            guard let handler = self.refreshUIHandler else { return }
            handler()
        }
    }

    /// Unsubscribes to vehicle odometer data.
    func unsubscribeToVehicleOdometer() {
        let unsubscribeToVehicleOdometer = SDLUnsubscribeVehicleData()
        unsubscribeToVehicleOdometer.odometer = true
        sdlManager.send(request: unsubscribeToVehicleOdometer) { (request, response, error) in
            guard let response = response, response.resultCode == .success else { return }
            self.resetOdometer()
        }
    }

    /// Notification containing the updated vehicle data.
    ///
    /// - Parameter notification: A SDLOnVehicleData notification
    func vehicleDataNotification(_ notification: SDLRPCNotificationNotification) {
        guard let handler = refreshUIHandler, let onVehicleData = notification.notification as? SDLOnVehicleData, let odometer = onVehicleData.odometer else {
            return
        }

        vehicleOdometerData = "\(VehicleDataOdometerName): \(odometer) km"
        handler()
    }

    /// Resets the odometer data
    fileprivate func resetOdometer() {
        vehicleOdometerData = "\(VehicleDataOdometerName): Unsubscribed"
    }
}

// MARK: - Get Vehicle Data

extension VehicleDataManager {
    /// Retreives the current vehicle data
    ///
    /// - Parameters:
    ///   - manager: The SDL manager
    ///   - triggerSource: Whether the menu item was selected by voice or touch
    ///   - parameter: The vehicle data to look for
    class func getAllVehicleData(with manager: SDLManager, triggerSource: SDLTriggerSource, vehicleDataType: String) {
        guard hasPermissionToAccessVehicleData(with: manager) else { return }

        SDLLog.d("App has permission to access vehicle data. Requesting all vehicle data...")
        let getAllVehicleData = SDLGetVehicleData(accelerationPedalPosition: true, airbagStatus: true, beltStatus: true, bodyInformation: true, clusterModeStatus: true, deviceStatus: true, driverBraking: true, eCallInfo: true, emergencyEvent: true, engineOilLife: true, engineTorque: true, externalTemperature: true, fuelLevel: true, fuelLevelState: true, fuelRange: true, gps: true, headLampStatus: true, instantFuelConsumption: true, myKey: true, odometer: true, prndl: true, rpm: true, speed: true, steeringWheelAngle: true, tirePressure: true, vin: true, wiperStatus: true)

        manager.send(request: getAllVehicleData) { (request, response, error) in
            guard didAccessVehicleDataSuccessfully(with: manager, response: response, error: error) else { return }

            var alertMessage = ""
            switch response!.resultCode {
            case .rejected:
                SDLLog.d("The request for vehicle data was rejected")
                alertMessage = "Rejected"
            case .disallowed:
                SDLLog.d("This app does not have the required permissions to access vehicle data")
                alertMessage = "Disallowed"
            case .success:
                if let vehicleData = response as? SDLGetVehicleDataResponse {
                    SDLLog.d("Request for vehicle data successful")
                    alertMessage = vehicleDataDescription(vehicleData, vehicleDataType: vehicleDataType)
                } else {
                    SDLLog.e("Request for vehicle data successful but no data returned")
                    alertMessage = "Unknown"
                }
            default: break
            }

            triggerSource == .menu ? manager.send(AlertManager.alertWithMessageAndCloseButton(alertMessage.isEmpty ? "No data available" : alertMessage)) : manager.send(SDLSpeak(tts: alertMessage))
        }
    }

    /// Returns a description of the vehicle data.
    ///
    /// - Parameters:
    ///   - vehicleData: All vehicle data
    ///   - vehicleDataType: The vehicle data to get a description of
    /// - Returns: A description of the vehicle data
    class func vehicleDataDescription(_ vehicleData: SDLGetVehicleDataResponse, vehicleDataType: String) -> String {
        let notAvailable = "Vehicle data not available"
        var message = "\(vehicleDataType): "
        switch vehicleDataType {
        case ACAccelerationPedalPositionMenuName:
            message += vehicleData.accPedalPosition?.description ?? notAvailable
        case ACAirbagStatusMenuName:
            message += vehicleData.airbagStatus?.description ?? notAvailable
        case ACBeltStatusMenuName:
            message += vehicleData.beltStatus?.description ?? notAvailable
        case ACBodyInformationMenuName:
            message += vehicleData.bodyInformation?.description ?? notAvailable
        case ACClusterModeStatusMenuName:
            message += vehicleData.clusterModeStatus?.description ?? notAvailable
        case ACDeviceStatusMenuName:
            message += vehicleData.deviceStatus?.description ?? notAvailable
        case ACDriverBrakingMenuName:
            message += vehicleData.driverBraking?.description ?? notAvailable
        case ACECallInfoMenuName:
            message += vehicleData.eCallInfo?.description ?? notAvailable
        case ACEmergencyEventMenuName:
            message += vehicleData.emergencyEvent?.description ?? notAvailable
        case ACEngineOilLifeMenuName:
            message += vehicleData.engineOilLife?.description ?? notAvailable
        case ACEngineTorqueMenuName:
            message += vehicleData.engineTorque?.description ?? notAvailable
        case ACExternalTemperatureMenuName:
            message += vehicleData.externalTemperature?.description ?? notAvailable
        case ACFuelLevelMenuName:
            message += vehicleData.fuelLevel?.description ?? notAvailable
        case ACFuelLevelStateMenuName:
            message += vehicleData.fuelLevel_State?.rawValue.rawValue ?? notAvailable
        case ACFuelRangeMenuName:
            message += vehicleData.fuelRange?.description ?? notAvailable
        case ACGPSMenuName:
            message += vehicleData.gps?.description ?? notAvailable
        case ACHeadLampStatusMenuName:
            message += vehicleData.headLampStatus?.description ?? notAvailable
        case ACInstantFuelConsumptionMenuName:
            message += vehicleData.instantFuelConsumption?.description ?? notAvailable
        case ACMyKeyMenuName:
            message += vehicleData.myKey?.description ?? notAvailable
        case ACOdometerMenuName:
            message += vehicleData.odometer?.description ?? notAvailable
        case ACPRNDLMenuName:
            message += vehicleData.prndl?.rawValue.rawValue ?? notAvailable
        case ACSpeedMenuName:
            message += vehicleData.speed?.description ?? notAvailable
        case ACSteeringWheelAngleMenuName:
            message += vehicleData.steeringWheelAngle?.description ?? notAvailable
        case ACTirePressureMenuName:
            message += vehicleData.tirePressure?.description ?? notAvailable
        case ACVINMenuName:
            message += vehicleData.vin?.description ?? notAvailable
        default: break
        }

        // Trim all non a-Z0-9 characters and truncate the string if it is more than 500 characters
        let trimmedString = message.filter { "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 :".contains($0) }
        let condensedString = String(trimmedString).condensedWhitespace
        return condensedString.trunc(length: 490)
    }

    /// Checks if the app has the required permissions to access vehicle data
    ///
    /// - Parameter manager: The SDL manager
    /// - Returns: True if the app has permission to access vehicle data, false if not
    class func hasPermissionToAccessVehicleData(with manager: SDLManager) -> Bool {
        SDLLog.d("Checking if app has permission to access vehicle data...")

        guard manager.permissionManager.isRPCAllowed("GetVehicleData") else {
            let alert = AlertManager.alertWithMessageAndCloseButton("This app does not have the required permissions to access vehicle data")
            manager.send(request: alert)
            return false
        }

        return true
    }

    /// Checks if Core sent back vehicle data
    ///
    /// - Parameters:
    ///   - manager: The SDL manager
    ///   - response: The response from Core
    ///   - error: An error from Core
    /// - Returns: True if Core sent back vehicle data, false if not
    class func didAccessVehicleDataSuccessfully(with manager:SDLManager, response: SDLRPCResponse?, error: Error?) -> Bool {
        SDLLog.d("Checking if Core returned vehicle data")

        guard response != nil, error == nil else {
            let alert = AlertManager.alertWithMessageAndCloseButton("Something went wrong while getting vehicle data")
            manager.send(request: alert)
            return false
        }

        return true
    }
}

// MARK: - Phone Calls

extension VehicleDataManager {
    /// Checks if the head unit has the ability and/or permissions to make a phone call. If it does, the phone number is dialed.
    ///
    /// - Parameter manager: The SDL manager
    /// - phoneNumber: A phone number to dial
    class func checkPhoneCallCapability(manager: SDLManager, phoneNumber: String) {
        SDLLog.d("Checking phone call capability")
        manager.systemCapabilityManager.updateCapabilityType(.phoneCall, completionHandler: { (error, systemCapabilityManager) in
            guard let phoneCapability = systemCapabilityManager.phoneCapability else {
                manager.send(AlertManager.alertWithMessageAndCloseButton("The head unit does not support the phone call capability"))
                return
            }
            if phoneCapability.dialNumberEnabled?.boolValue ?? false {
                SDLLog.d("Dialing phone number \(phoneNumber)...")
                dialPhoneNumber(phoneNumber, manager: manager)
            } else {
                manager.send(AlertManager.alertWithMessageAndCloseButton("A phone call can not be made"))
            }
        })
    }

    /// Dials a phone number.
    ///
    /// - Parameters:
    ///   - phoneNumber: A phone number to dial
    ///   - manager: The SDL manager
    private class func dialPhoneNumber(_ phoneNumber: String, manager: SDLManager) {
        let dialNumber = SDLDialNumber(number: phoneNumber)
        manager.send(request: dialNumber) { (requst, response, error) in
            guard let success = response?.resultCode else { return }
            SDLLog.d("Sent dial number request: \(success == .success ? "successfully" : "unsuccessfully").")
        }
    }
}

extension String {
    var condensedWhitespace: String {
        let components = self.components(separatedBy: NSCharacterSet.whitespacesAndNewlines)
        return components.filter { !$0.isEmpty }.joined(separator: " ")
    }

    func trunc(length: Int, trailing: String = "…") -> String {
        return (self.count > length) ? self.prefix(length) + trailing : self
    }
}

extension SDLVehicleDataEventStatus {
    var description: String {
        switch self {
        case .noEvent:
            return "no event"
        case .no:
            return "no"
        case .yes:
            return "yes"
        case .notSupported:
            return "not supported"
        case .fault:
            return "fault"
        default: break
        }
        return ""
    }
}
